src/app/api/upload/route.ts

```diff
@@ -0,0 +1,98 @@
+import { md5 } from "@/lib/md5";
+import { apiError, concat } from "@/lib/utils";
+import withUser from "@/lib/with-user";
+import { NextResponse } from "next/server";
+import { ok, err } from "@justmiracle/result";
+import { R2 } from "@/lib/r2";
+import { db } from "@/db";
+import { user as userTable, user_file } from "@/db/schema";
+import { eq } from "drizzle-orm";
+import { generateId } from "lucia";
+
+// 500MB
+const STORAGE_LIMIT = 500 * 1024 * 1024;
+
+export const POST = withUser(async ({ req, user }) => {
+  const formData = await req.formData().then(ok).catch(err);
+
+  if (formData.error) {
+    return apiError({
+      message: "Something went wrong, please try again later",
+      detailedMessage: formData.error.message,
+      status: 500,
+    });
+  }
+
+  const file = formData.value.get("file");
+
+  if (!(file instanceof File)) {
+    return apiError({
+      message: "Invalid file",
+      status: 400,
+    });
+  }
+
+  const newStorageUsage = user.storageUsage + file.size;
```

Comment:
This will have problem with some race condition. For example:

Let say user A has used 500 bytes. He/she upload 2 files at the same times. One is 300 bytes, and another is 100 bytes.

Depends on the order of execute. The storage usage will either increased by 300, 100 or 400.

Reply:
Done